### ISSUE TITLE
[ONNX] Internal assertion failure when export gpt2 model with `export_modules_as_functions=True`

### ISSUE THREAD

**Issue by HaoboGu** [pytorch/pytorch#79839]
### 🐛 Describe the bug

Hello, I'm using `torch.onnx.export` to convert my gpt2 model to onnx format. But when I set `        opset_version=15` and `export_modules_as_functions=True`, the program reports an error:
```
Traceback (most recent call last):
  File "/Users/haobogu/Projects/python/cosine/.venv/lib/python3.7/site-packages/torch/onnx/__init__.py", line 309, in export
    export_modules_as_functions)
  File "/Users/haobogu/Projects/python/cosine/.venv/lib/python3.7/site-packages/torch/onnx/utils.py", line 122, in export
    custom_opsets=custom_opsets, export_modules_as_functions=export_modules_as_functions)
  File "/Users/haobogu/Projects/python/cosine/.venv/lib/python3.7/site-packages/torch/onnx/utils.py", line 736, in _export
    graph, export_modules_as_functions, list(params_dict.keys()))
RuntimeError: ns_a[i]->kind() == ns_b[i]->kind()
INTERNAL ASSERT FAILED at "/Users/distiller/project/pytorch/torch/csrc/jit/passes/onnx/function_extraction.cpp":146, please report a bug to PyTorch.
```

### Versions

Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A
OS: macOS 12.4 (x86_64)
GCC version: Could not collect
Clang version: 13.0.1
CMake version: version 3.22.3
Libc version: N/A
Python version: 3.7.9 (v3.7.9:13c94747c7, Aug 15 2020, 01:31:08)  [Clang 6.0 (clang-600.0.57)] (64-bit runtime)
Python platform: Darwin-21.5.0-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Versions of relevant libraries:
[pip3] numpy==1.21.6
[pip3] torch==1.11.0
[pip3] torchvision==0.12.0
[conda] Could not collect

**Comment by justinchuby** [comment]
Does `export_modules_as_functions=False` give you the same error?

**Comment by HaoboGu** [comment]
@justinchuby No, setting `export_modules_as_functions=False` removes this error

**Comment by justinchuby** [comment]
Could you share a script that will help us reproduce the error?

**Comment by HaoboGu** [comment]
@justinchuby here is the script

```python
from transformers import GPT2Config
from transformers.models.gpt2.modeling_gpt2 import GPT2LMHeadModel
import torch

def convert_gpt2_model_to_onnx() -> str:
    """
    convert pytorch model to onnx format
    """
    # Load model and set the model to eval mode
    model: GPT2LMHeadModel = GPT2LMHeadModel.from_pretrained('sshleifer/tiny-gpt2')
    model.eval()

    # batch_size, input_ids_length and past_sequence_length are dynamic axes
    # We have to initialize a random input(the value doesn't matter) for the model, because the converting requires execution of the model
    # See: https://pytorch.org/tutorials/advanced/super_resolution_with_onnxruntime.html
    batch_size = 3
    input_ids_length = 2
    past_sequence_length = 1
    config: GPT2Config = model.config
    num_attention_heads = config.n_head
    hidden_size = config.n_embd
    num_layer = config.n_layer
    vocab_size = config.vocab_size
    config.is_decoder = True

    # past(`past_key_values` in model) is a list, its length is num_layer.
    # each element in the list is a tuple(key, value), and key/value's shape is past_shape,
    # aka [batch_size, n_heads, past_sequence_length, embd_size_each_head]
    past_shape = [batch_size, num_attention_heads,
                  past_sequence_length, int(hidden_size/num_attention_heads)]
    past = [(torch.rand(past_shape, dtype=torch.float32, device='cpu'), torch.rand(past_shape, dtype=torch.float32, device='cpu'))
            for _ in range(num_layer)]

    # input_ids is a [batch_length, input_ids_length] tensor
    input_ids = torch.randint(
        low=0,
        high=vocab_size - 1,
        size=(batch_size, input_ids_length),
        dtype=torch.long,
        device='cpu',
    )

    # attention_mask is a 0/1 tensor of [batch_size, past_sequence_length + input_ids_length]
    attention_mask = torch.ones(
        [batch_size, past_sequence_length + input_ids_length]).to(torch.long)

    # token_type_ids is not needed in our case, its size is [batch_size, input_ids_length]
    token_type_ids = torch.zeros([batch_size, input_ids_length]).to(torch.long)

    # position_ids, size is [batch_size, input_ids_length]
    position_ids = attention_mask.long().cumsum(-1) - 1
    position_ids.masked_fill_(position_ids < 0, 0)
    position_ids = position_ids[:, past_sequence_length:].to(torch.long)

    # Run the model and get output from the model
    output = model(input_ids, past_key_values=past, attention_mask=attention_mask,
                   position_ids=position_ids, return_dict=True, use_cache=True)

    # Set output names
    output_names = ['logits']
    for i in range(num_layer):
        output_names.append("present_key" + str(i))
        output_names.append("present_value" + str(i))

    # Set input_names
    input_names = ['input_ids']
    for i in range(num_layer):
        input_names.append("past_key" + str(i))
        input_names.append("past_value" + str(i))
    input_names += ['attention_mask', 'position_ids']

    # Set dynamic axes
    dynamic_axes = {}
    dynamic_axes['input_ids'] = {0: 'batch_size', 1: 'input_ids_length'}
    dynamic_axes['attention_mask'] = {0: 'batch_size', 1: 'total_length'}
    dynamic_axes['position_ids'] = {0: 'batch_size', 1: 'input_ids_length'}
    dynamic_axes['logits'] = {0: 'batch_size', 1: 'input_ids_length'}
    for i in range(num_layer):
        dynamic_axes['past_key' +
                     str(i)] = {0: 'batch_size', 2: 'past_sequence_length'}
        dynamic_axes['past_value' +
                     str(i)] = {0: 'batch_size', 2: 'past_sequence_length'}
        dynamic_axes['present_key' +
                     str(i)] = {0: 'batch_size', 2: 'total_length'}
        dynamic_axes['present_value' +
                     str(i)] = {0: 'batch_size', 2: 'total_length'}

    # The first input is required, and other inputs can be passed to torch.onnx.export using dict
    inputs = (input_ids, {
        'attention_mask': attention_mask,
        'position_ids': position_ids,
        'past_key_values': past,
    })

    # Do export using torch.onnx.export
    exported_model = "converted_model.onnx"
    torch.onnx.export(
        model,
        args=inputs,
        f=exported_model,
        export_params=True,
        verbose=False,
        input_names=input_names,
        output_names=output_names,
        dynamic_axes=dynamic_axes,
        opset_version=15,
        export_modules_as_functions=True
    )
    return exported_model

convert_gpt2_model_to_onnx()
```

**Comment by HaoboGu** [comment]
@justinchuby any ideas?

**Comment by justinchuby** [comment]
Thanks for the script. We are tracking a few instabilities with export_modules_as_functions. Will update here

**Comment by justinchuby** [comment]
cc @BowenBao

**Comment by BowenBao** [comment]
Fixed by #80999
